File: javascript/Maths/Mandelbrot.js
export function getRGBData(
  imageWidth = 800,
  imageHeight = 600,
  figureCenterX = -0.6,
  figureCenterY = 0,
  figureWidth = 3.2,
  maxStep = 50,
  useDistanceColorCoding = true
) {
  if (imageWidth <= 0) {
    throw new Error('imageWidth should be greater than zero')
  }

  if (imageHeight <= 0) {
    throw new Error('imageHeight should be greater than zero')
  }

  if (maxStep <= 0) {
    throw new Error('maxStep should be greater than zero')
  }

  const rgbData = []
  const figureHeight = (figureWidth / imageWidth) * imageHeight

  for (let imageX = 0; imageX < imageWidth; imageX++) {
    rgbData[imageX] = []
    for (let imageY = 0; imageY < imageHeight; imageY++) {
      
      const figureX = figureCenterX + (imageX / imageWidth - 0.5) * figureWidth
      const figureY =
        figureCenterY + (imageY / imageHeight - 0.5) * figureHeight

      const distance = getDistance(figureX, figureY, maxStep)

      rgbData[imageX][imageY] = useDistanceColorCoding
        ? colorCodedColorMap(distance)
        : blackAndWhiteColorMap(distance)
    }
  }

  return rgbData
}

function blackAndWhiteColorMap(distance) {
  return distance >= 1 ? [0, 0, 0] : [255, 255, 255]
}

function colorCodedColorMap(distance) {
  if (distance >= 1) {
    return [0, 0, 0]
  } else {
    
    const hue = 360 * distance
    const saturation = 1
    const val = 255
    const hi = Math.floor(hue / 60) % 6
    const f = hue / 60 - Math.floor(hue / 60)

    const v = val
    const p = 0
    const q = Math.floor(val * (1 - f * saturation))
    const t = Math.floor(val * (1 - (1 - f) * saturation))

    switch (hi) {
      case 0:
        return [v, t, p]
      case 1:
        return [q, v, p]
      case 2:
        return [p, v, t]
      case 3:
        return [p, q, v]
      case 4:
        return [t, p, v]
      default:
        return [v, p, q]
    }
  }
}

function getDistance(figureX, figureY, maxStep) {
  let a = figureX
  let b = figureY
  let currentStep = 0
  for (let step = 0; step < maxStep; step++) {
    currentStep = step
    const aNew = a * a - b * b + figureX
    b = 2 * a * b + figureY
    a = aNew

    if (a * a + b * b > 4) {
      break
    }
  }
  return currentStep / (maxStep - 1)
}

File: csharp/Algorithms/Other/Mandelbrot.cs
using System;
using SkiaSharp;

namespace Algorithms.Other;

public static class Mandelbrot
{
    private const byte Alpha = 255;

    public static SKBitmap GetBitmap(
        int bitmapWidth = 800,
        int bitmapHeight = 600,
        double figureCenterX = -0.6,
        double figureCenterY = 0,
        double figureWidth = 3.2,
        int maxStep = 50,
        bool useDistanceColorCoding = true)
    {
        if (bitmapWidth <= 0)
        {
            throw new ArgumentOutOfRangeException(
                nameof(bitmapWidth),
                $"{nameof(bitmapWidth)} should be greater than zero");
        }

        if (bitmapHeight <= 0)
        {
            throw new ArgumentOutOfRangeException(
                nameof(bitmapHeight),
                $"{nameof(bitmapHeight)} should be greater than zero");
        }

        if (maxStep <= 0)
        {
            throw new ArgumentOutOfRangeException(
                nameof(maxStep),
                $"{nameof(maxStep)} should be greater than zero");
        }

        var bitmap = new SKBitmap(bitmapWidth, bitmapHeight);
        var figureHeight = figureWidth / bitmapWidth * bitmapHeight;

        for (var bitmapX = 0; bitmapX < bitmapWidth; bitmapX++)
        {
            for (var bitmapY = 0; bitmapY < bitmapHeight; bitmapY++)
            {
                
                var figureX = figureCenterX + ((double)bitmapX / bitmapWidth - 0.5) * figureWidth;
                var figureY = figureCenterY + ((double)bitmapY / bitmapHeight - 0.5) * figureHeight;

                var distance = GetDistance(figureX, figureY, maxStep);

                bitmap.SetPixel(
                    bitmapX,
                    bitmapY,
                    useDistanceColorCoding ? ColorCodedColorMap(distance) : BlackAndWhiteColorMap(distance));
            }
        }

        return bitmap;
    }

    private static SKColor BlackAndWhiteColorMap(double distance) =>
        distance >= 1
            ? new SKColor(0, 0, 0, Alpha)
            : new SKColor(255, 255, 255, Alpha);

    private static SKColor ColorCodedColorMap(double distance)
    {
        if (distance >= 1)
        {
            return new SKColor(0, 0, 0, Alpha);
        }

        var hue = 360 * distance;
        double saturation = 1;
        double val = 255;
        var hi = (int)Math.Floor(hue / 60) % 6;
        var f = hue / 60 - Math.Floor(hue / 60);

        var v = (byte)val;
        const byte p = 0;
        var q = (byte)(val * (1 - f * saturation));
        var t = (byte)(val * (1 - (1 - f) * saturation));

        switch (hi)
        {
            case 0: return new SKColor(v, t, p, Alpha);
            case 1: return new SKColor(q, v, p, Alpha);
            case 2: return new SKColor(p, v, t, Alpha);
            case 3: return new SKColor(p, q, v, Alpha);
            case 4: return new SKColor(t, p, v, Alpha);
            default: return new SKColor(v, p, q, Alpha);
        }
    }

    private static double GetDistance(double figureX, double figureY, int maxStep)
    {
        var a = figureX;
        var b = figureY;
        var currentStep = 0;
        for (var step = 0; step < maxStep; step++)
        {
            currentStep = step;
            var aNew = a * a - b * b + figureX;
            b = 2 * a * b + figureY;
            a = aNew;

            if (a * a + b * b > 4)
            {
                break;
            }
        }

        return (double)currentStep / (maxStep - 1);
    }
}
